panel/widgets/chatbox.py

```diff
@@ -0,0 +1,359 @@
+                margin=margin,
+                styles={"color": "grey"},
+                align=horizontal_align,
+            )
+            row = Column(name, row, **container_params)
+
+        self._composite[:] = [row]
+
+
+class ChatBox(CompositeWidget):
+    value = param.List(
+        doc="""List of messages, mapping user to message,
+        e.g. `[{'You': 'Welcome!'}]`""",
+        item_type=Dict,
+        default=[],
+    )
+
+    primary_user = param.String(
+        doc="""
+            Name of the primary input user;
+            the first key found in value
+            will be used if unspecified
+        """,
+        default=None,
+    )
+
+    allow_input = param.Boolean(
+        doc="""
+            Whether to allow the primary user to interactively
+            enter messages.
+        """,
+        default=True,
+    )
+
+    user_icons = param.Dict(
+        doc="""Dictionary mapping name of users to their icons,
+        e.g. `[{'You': 'path/to/icon.png'}]`""",
+        default={},
+    )
+
+    user_colors = param.Dict(
+        doc="""Dictionary mapping name of users to their colors,
+        e.g. `[{'You': 'red'}]`""",
+        default={},
+    )
+
+    _composite_type: ClassVar[Type[ListPanel]] = Column
+
+    def __init__(self, **params):
+        # set up parameters
+        if params.get("width") and params.get("height") and "sizing_mode" not in params:
+            params["sizing_mode"] = None
+
+        super().__init__(**params)
+
+        # Set up layout
+        layout = {
+            p: getattr(self, p)
+            for p in Layoutable.param
+            if p not in ("name", "height", "margin") and getattr(self, p) is not None
+        }
+        chat_layout = dict(
+            sizing_mode="stretch_both",
+            height=None,
+            margin=0,
+            **layout,
+        )
+        self._chat_title = StaticText(
+            value=f"{self.name}",
+            styles={"font-size": "1.5em"},
+            align="center",
+        )
+        self._chat_log = Column(scroll=True, **chat_layout)
+
+        box_objects = [self._chat_title] if self.name else []
+        box_objects.append(self._chat_log)
+        if self.allow_input:
+            self._input_message = TextInput(
+                name="Type your message",
+                placeholder="Press Enter to send",
+                sizing_mode="stretch_width",
+                margin=0,
+            )
+            self._input_message.param.watch(self._enter_message, "value")
+            box_objects.extend([VSpacer(height_policy="min"), self._input_message])
+        self._composite[:] = box_objects
+
+        # add interactivity
+        self.param.watch(self._refresh_log, "value")
+        self.param.trigger("value")
+
+
+    def _generate_dark_color(self) -> str:
+        """
+        Generate a random dark color in hexadecimal format.
+        """
+        r, g, b = random.randint(0, 127), random.randint(0, 127), random.randint(0, 127)
```

Comment:
Maybe this could use the username to generate the color, so it is the same color for the same username each time you instantiate the chatbox.

Reply:
Brilliant idea!